### DISCUSSION TITLE
linux sample can not achieve 0-rtt?

### DISCUSSION THREAD

**Discussion by muyiwq** [microsoft/msquic#1771] [Q&A] [answered]
System: centos 7
In the second connection in the sample
1. Specify -ticket:resumesession to successfully reach this part of the logic
2. 
Settings.ServerResumptionLevel =QUIC_SERVER_RESUME_AND_ZERORTT;
Settings.IsSet.ServerResumptionLevel = TRUE;
3.StreamSend(QUIC_SEND_FLAG_ALLOW_0_RTT | QUIC_SEND_FLAG_FIN);

Why is there no 0-rtt packet captured by tcpdump, but it is still an Initial packet. Do I need to do anything else?
Can linux support 0-rtt? How to carry application data in 0-rtt?

**Comment by nibanks** [comment, marked as answer]
Currently, in the sample, the client doesn't open and send on a stream until after the handshake. By then, you already have the 1-RTT keys so you don't use 0-RTT. If you want to try to exercise 0-RTT, you'd need to move `ClientSend` to **before** the call to `ConnectionStart` (along with the changes you list above).

**Comment by muyiwq** [comment]
Thank you very much for your reply! wish you a happy life!
centos 7


1.0-rtt has been caught, I would like to ask if there is any way to get ssl_key. Stream and data can be seen in wireshark?


2. How does RunClient(EQUICESULT *eRes) obtain the success or failure of streaming data?


Since ClientSend (EQUICRESULT *eRes) is sent asynchronously, I assigned a value to eWriteResult in the QUIC_STREAM_EVENT_SEND_COMPLETE event of ClientStreamCallback (EQUICRESULT *eRes), but I did not get the sending result at the end of the RunClient function. Only after RegistrationClose() can the result be obtained,


I now want to send RunClient(EQUICESULT *eRes) in multiple threads, and it will not execute to the RegistrationClose() function.
How can I know where ClientSend has successfully sent data asynchronously as soon as possible?





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "microsoft/msquic"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年6月25日(星期五) 晚上11:18
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [microsoft/msquic] linux sample can not achieve 0-rtt? (#1771)





 
Currently, in the sample, the client doesn't open and send on a stream until after the handshake. By then, you already have the 1-RTT keys so you don't use 0-RTT. If you want to try to exercise 0-RTT, you'd need to move ClientSend to before the call to ConnectionStart (along with the changes you list above).

**Reply by nibanks** [reply]
It's happening after the call to `RegistrationClose` because nothing else waits on the main thread. The client calls `ConnectionStart` and then simply waits. That's not a bug, it's by design.

**Reply by nibanks** [reply]
To get the SSL secrets, you'd need to wire up [QUIC_PARAM_CONN_TLS_SECRETS](https://github.com/microsoft/msquic/blob/main/src/inc/msquicp.h#L135).

**Reply by muyiwq** [reply]
centos 7
Neither client nor the server can get the key in QUIC_PARAM_CONN_TLS_SECRETS.

After QUIC_CONNECTION_EVENT_CONNECTED or before ConnectionStart, both return QUIC_STATUS_NOT_SUPPORTED
              QUIC_STATUS Status =
                  MsQuic->SetParam(
                      Connection,
                      QUIC_PARAM_LEVEL_CONNECTION,
                      QUIC_PARAM_CONN_TLS_SECRETS,
                      sizeof(TlsSecrets),
                      (uint8_t*)&TlsSecrets);
The failure status is QUIC_STATUS_NOT_SUPPORTED, It means I did not define CXPLAT_TLS_SECRETS_SUPPORT

I have defined this macro in the code or in the compile option, and it still returns QUIC_STATUS_NOT_SUPPORTED status.
Where do I need to set SUPPORTED setting?

WriteSslKeyLogFile(SslKeyLogFile, TlsSecrets)
Write to the file I specify

Thanks again
